### PR TITLE
Multiple geth v2

### PR DESCRIPTION
**Pull request by markya0616**

### CODE REVIEW

File: service/indexer/indexer_test.go [hunk 1]
```diff
@@ -810,34 +901,35 @@ var _ = Describe("Indexer Test", func() {
...
 			}()
 
-			err := idx.Listen(ctx, ch, 0)
+			err := idx.Listen(ctx, 0)
 			Expect(err).Should(Equal(context.Canceled))
 		})
 	})
 
-	Context("Listen() old block with Reorg", func() {
+	Context("Listen() reorg with older header", func() {
```
Comment (kimmylin) [inline service/indexer/indexer_test.go]: add a test that have headers from different eth client?

File: cmd/indexer/root.go [hunk 2]
```diff
@@ -92,12 +90,17 @@ var ServerCmd = &cobra.Command{
...
+		for _, c := range ethClients {
+			ethClient, err := NewEthConn(c)
+			if err != nil {
+				log.Error("Failed to new a eth client", "err", err)
+				return err
+			}
+			defer ethClient.Close()
+
+			clients = append(clients, ethClient)
 		}
```
[the remark this answers is not in the document]
Comment (markya0616) [reply cmd/indexer/root.go]: OK

File: service/indexer/indexer.go [hunk 3]
```diff
@@ -83,38 +92,85 @@ func (idx *indexer) SubscribeErc20Tokens(ctx context.Context, addresses []string
...
+		case <-childCtx.Done():
+			err := childCtx.Err()
+			log.Warn("Receive subscribe context error", "client", index, "err", err)
+			return err
+		}
+	}
+}
+
+func (idx *indexer) Listen(ctx context.Context, fromBlock int64) error {
+	outChannel := make(chan *Result, 100)
```
[the remark this answers is not in the document]
Comment (markya0616) [reply service/indexer/indexer.go]: OK. 50 * len(clients)

File: service/indexer/indexer.go [hunk 4]
```diff
@@ -83,38 +92,85 @@ func (idx *indexer) SubscribeErc20Tokens(ctx context.Context, addresses []string
...
+		log.Warn("Failed to subscribe event for new header", "client", index, "err", err)
 		return err
 	}
 
 	for {
 		select {
 		case head := <-ch:
-			if fromBlock > head.Number.Int64() {
-				log.Trace("Ignore old header", "number", head.Number, "hash", head.Hash().Hex())
+			outChannel <- &Result{
```
[the remark this answers is not in the document]
Comment (markya0616) [reply service/indexer/indexer.go]: Will when we got result from out channel

File: service/indexer/indexer.go [hunk 5]
```diff
@@ -175,70 +175,45 @@ func (idx *indexer) Listen(ctx context.Context, fromBlock int64) error {
...
-		}
-	}
-
-	// Ensure the from block is lager than current block
-	if from-1 > header.Number {
-		block, err := idx.latestClient.BlockByNumber(ctx, big.NewInt(from-1))
-		if err != nil {
-			log.Error("Failed to get block", "number", from, "err", err)
-			return nil, nil, err
+			log.Warn("The latest header doesn't existed", "from", from)
```
[the remark this answers is not in the document]
Comment (markya0616) [reply service/indexer/indexer.go]: OK

File: service/indexer/indexer.go [hunk 6]
```diff
@@ -83,106 +94,127 @@ func (idx *indexer) SubscribeErc20Tokens(ctx context.Context, addresses []string
...
+				time.Sleep(retrySubscribeTime)
+			}
+		}(listenCtx, outChannel, i)
+	}
+	// single thread to process the messages from channel in sequential
+	for {
+		select {
+		case result := <-outChannel:
+			header := result.header
+			if fromBlock > header.Number.Int64() {
```
[the remark this answers is not in the document]
Comment (markya0616) [reply service/indexer/indexer.go]: OK

File: service/indexer/indexer.go [hunk 7]
```diff
@@ -271,74 +302,84 @@ func (idx *indexer) insertBlocks(ctx context.Context, blocks []*types.Block, reo
...
+	// 2. on the same chain, we don't need to reorg
 	var blocksToInsert []*types.Block
-	if target == 0 || bytes.Equal(block.ParentHash().Bytes(), idx.currentHeader.Hash) {
+	if idx.currentHeader.Number+1 == header.Number.Int64() && bytes.Equal(block.ParentHash().Bytes(), idx.currentHeader.Hash) {
 		blocksToInsert = append(blocksToInsert, block)
 		return idx.insertBlocks(ctx, blocksToInsert, nil)
 	}
 
+	// Find targetTD to check whether we need to handle it
+	targetTD, err := idx.getTd(ctx, header.Hash().Bytes())
```
[the remark this answers is not in the document]
Comment (markya0616) [reply service/indexer/indexer.go]: OK

File: service/indexer/indexer_test.go (hunk 1 above)
Comment (markya0616) [reply service/indexer/indexer_test.go]: OK

File: service/indexer/indexer.go [hunk 8]
```diff
@@ -177,37 +184,51 @@ func (idx *indexer) Listen(ctx context.Context, fromBlock int64) error {
...
 }
 
-func (idx *indexer) getLocalState(ctx context.Context, from int64) (*model.Header, *big.Int, error) {
+// loadLocalState loads the local state from DB and sets current header and TD.
+// If there is no state in DB, get the from block from Ethereum and insert it.
+func (idx *indexer) loadLocalState(ctx context.Context, from int64) error {
 	// Get latest header from db
 	header, err := idx.manager.LatestHeader()
 	if err != nil {
 		if common.NotFoundError(err) {
```
Comment (kimmylin) [inline service/indexer/indexer.go]: combine these two if stmts

**Review by countvonzero — APPROVED** [review]

**Review by Elenore23 — APPROVED** [review]

**Pull request merged**